### PR TITLE
Fix zero-grad behavior when entering the validation loop

### LINKED ISSUE #18713: Should the model's grads be cleared before entering the validation loop?
### Discussion

Our trainer performs optimization in this order: 1) `zero_grad()` 2) `backward()` 3) `optimizer.step()`
This means if validation is run after such an optimization step (either the last one in the epoch or according to the Trainer's `val_check_interval` during the epoch), gradients would still be around. 
To minimize the memory usage and risk of OOM during validation, we [`zero_grad` the model params ](https://github.com/Lightning-AI/lightning/blob/343f80436d3233bac155e77b92891f6281ff90d1/src/lightning/pytorch/loops/evaluation_loop.py#L244)before entering the validation loop. 

In normal use cases, this `zero_grad` call is not noticeable. NeMo's MLPerf team reported however that

> touching the grads there is causing some of the all gathers we are doing  before the validation_step to get exposed and not allowing it to be overlapped with the compute of validation_step. We want the all_gathers of all the buckets other than first to be overlapped with the val_step.

**Option 1:** Remove the explicit `zero_grad` call as suggested by NeMo.
If this `zero_grad` call were to be removed, the implication is that grads will be available during validation (i.e. `all(p.grad is not None for p in self.parameters())` and potentially use more memory than expected. 

**Option 2:** Change the optimization step
An alternative would be to change the order to 1) `backward()` 3) `optimizer.step()` 3) `zero_grad()`. Then, an explicit zero-ing at the beginning of validation wouldn't be necessary. But this would potentially be a breaking change for some users who rely on this order and have overridden hooks accordingly.

**Option 3:** Make it possible to override the `zero_grad` call (turning it off) in the validation loop. For example, in the form of a hook `LightningModule.on_evaluation_zero_grad()` similar to our `on_validation_model_eval()` hooks.

cc @borda

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #18713

This PR does the following:

1. Fixes an issue in which the validation loop would erase gradients when validation happened in the middle of a gradient accumulation phase (see the added test).
2. The `zero_grad()` call is now wrapped by a new hook `LightningModule.on_validation_model_zero_grad()` that the user can override if they don't want the default behavior (#18713).
3. We still ensure that we clear the gradients before entering any of the trainer stages `trainer.validate/test/predict` as a precaution (no change there). 

----
:books: Documentation preview :books:: https://pytorch-lightning--18710.org.readthedocs.build/en/18710/

cc @carmocca @awaelchli @borda @justusschock

### CODE REVIEW

File: src/lightning/pytorch/trainer/trainer.py [hunk 1]
```diff
@@ -1023,6 +1023,9 @@ def _run_stage(self) -> Optional[Union[_PREDICT_OUTPUT, _EVALUATE_OUTPUT]]:
         # wait for all to join if on distributed
         self.strategy.barrier("run-stage")
 
+        # TODO: Remove `set_to_none=True` when PyTorch 2.0 is the minimum supported version
+        self.lightning_module.zero_grad(set_to_none=True)
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/trainer.py:1027]: nit: one easy way to avoid having to remember about this TODO is to condition the call based on `_TORCH_GREATER_EQUAL_2_0`,

File: tests/tests_pytorch/loops/test_loops.py [hunk 2]
```diff
@@ -851,3 +851,34 @@ def _get_iterator(self):
             3,  # teardown on epoch 2, workers from epoch 2 get destroyed
         ]
     assert val_dataloader.shutdown_workers_epochs == expected
+
+
+def test_validation_during_gradient_accumulation_window(tmp_path):
```
Comment (awaelchli) [inline tests/tests_pytorch/loops/test_loops.py:856]: Test fails on master

File: src/lightning/pytorch/core/hooks.py [hunk 3]
```diff
@@ -151,6 +152,13 @@ def on_predict_batch_end(self, outputs: Optional[Any], batch: Any, batch_idx: in
 
         """
 
+    def on_validation_model_zero_grad(self) -> None:
+        """Called by the training loop to release gradients before entering the validation loop."""
+        if _TORCH_GREATER_EQUAL_2_0:
+            self.zero_grad()
+        else:
+            self.zero_grad(set_to_none=True)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/core/hooks.py:160]: That wouldn't be the same as `model.zero_grad()`. The loop never called optimizer's `zero_grad`. The motivation of the hook is to provide the override for specifically only the call that happens at the transition from training to validation. If you did an override of `LightningModule.zero_grad()`, you'd have to add logic to check whether you're in validation.

[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/pytorch/core/hooks.py:160]: Oh sorry, I mixed the two methods.

>  If you did an override of LightningModule.zero_grad(), you'd have to add logic to check whether you're in validation.

This is the point I was trying to make. Wouldn't it be fine to suggest this? Checking whether you're in validation is simple since `self` has a trainer reference

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/core/hooks.py:160]: I wanted to bring this in line with the other hooks we have for `model.eval()` at the training-validation boundary. This way, the user wouldn't override `nn.Module's zero_grad()` that applies more generally to the model rather than the training loop. A clear separation between Trainer hook and model method would be desireable I think.

[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/pytorch/core/hooks.py:160]: But `zero_grad` would be overridden in the `LightningModule`, not the underlying `nn.Module`. The LM already defines how the trainer interacts with the underlying module so from my point of view, it fits well.

It's fine if you still want the new hook, I just thought that we could avoid having one more hook to remember about

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/core/hooks.py:160]: I'd like to keep it unless there are stronger arguments against it.

I see two clear benefits in doing it this way. 1) The `zero_grad()` method remains untouched and can be used on the LightningModule outside the trainer context 2) Overriding the hook is cleaner because the alternative would mean the user would have to None-check `self.trainer` and check the trainer stage for validation and then also shouldn't forget to all super().zero_grad().

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/core/hooks.py (hunk 3 above)
Comment (Borda) [inline src/lightning/pytorch/core/hooks.py:160]:
```suggestion
        zero_grad_kw = {} if _TORCH_GREATER_EQUAL_2_0 else dict(set_to_none=True)
        self.zero_grad(**zero_grad_kw)
```

File: src/lightning/pytorch/trainer/trainer.py [hunk 4]
```diff
@@ -1023,6 +1023,11 @@ def _run_stage(self) -> Optional[Union[_PREDICT_OUTPUT, _EVALUATE_OUTPUT]]:
         # wait for all to join if on distributed
         self.strategy.barrier("run-stage")
 
+        if _TORCH_GREATER_EQUAL_2_0:
+            self.lightning_module.zero_grad()
+        else:
+            self.lightning_module.zero_grad(set_to_none=True)
```
Comment (Borda) [inline src/lightning/pytorch/trainer/trainer.py:1029]:
```suggestion
        zero_grad_kw = {} if _TORCH_GREATER_EQUAL_2_0 else dict(set_to_none=True)
        self.lightning_module.zero_grad(**zero_grad_kw)
```

**Review by Borda — APPROVED** [review]
just minor formating, feel free to ignore
